### ISSUE TITLE
Docker environment setup on EKS

### ISSUE THREAD

**Issue by ndenStanford** [aws/aws-neuron-sdk#361]
AWS newbie here. I am trying to update the neuron driver on the application deployed EKS instance. I am following the steps listed on [here](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/neuron-deploy/v1/containers-migration.html).  One of the steps mentioned is to perform docker environment setup [here](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/neuron-deploy/tutorials/tutorial-docker-env-setup.html#tutorial-docker-env-setup). I performed the mentioned steps on a **separate** EC2 machine, built the docker image, and pushed it to the ECR repo. I then deployed the pod on another EKS cluster. Once I interact with the deployed cluster via HTTPS request, I am seeing the following error in the log.  

2021-11-08 22:21:25,025 [WARN ] W-9003-sent_neuron_b6_1.0-stderr MODEL_LOG - [E neuron_runtime.cpp:683] **The Neuron Runtime GRPC server at address unix:/run/neuron.sock is unavailable. Is neuron-rtd running? Is socket /run/neuron.sock writable?**
2021-11-08 22:21:25,025 [WARN ] W-9003-sent_neuron_b6_1.0-stderr MODEL_LOG - /opt/workspace/KaenaPyTorchRuntime/neuron_op/neuron_op_impl.cpp:52: UserWarning: Neuron runtime cannot be initialized; falling back to CPU execution
2021-11-08 22:21:25,025 [WARN ] W-9003-sent_neuron_b6_1.0-stderr MODEL_LOG - Tensor output are ** NOT CALCULATED ** during CPU execution and only indicate tensor shape
2021-11-08 22:21:25,025 [INFO ] W-9003-sent_neuron_b6_1.0-stdout MODEL_LOG - Total Time in milliseconds = 10.934
2021-11-08 22:21:25,026 [INFO ] W-9003-sent_neuron_b6_1.0-stdout MODEL_LOG - Average Time per sentence = 10.934
2021-11-08 22:21:25,026 [INFO ] W-9003-sent_neuron_b6_1.0-stdout MODEL_LOG - Prediction appears to be from CPU, returning none.

Would anyone be able to shed some light on what might be causing the issue? I am thinking that it has to do with the neuron runtime installation in the perform docker environment setup [here](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/neuron-deploy/tutorials/tutorial-docker-env-setup.html#tutorial-docker-env-setup). Since I followed these steps on a separate EC2 machine when building the image, it is not being registered on the EKS cluster that I used in the deployment. With that said, I do not see how to follow those steps on the EKS cluster the same way I do for an EC2 machine (which I could access via SSH). Any help is greatly appreciated. Thank you!

**Comment by awsryg** [comment]
Hi ndenStanford, Latest Neuron no longer requires the use of neuron-rtd, which means that sidecar container of your previous yaml could be removed.  The driver must be updated on the host ami. The other components in the application container can be updated as well once the driver updates is performed on your cluster. Hope this helps.

**Comment by ndenStanford** [comment]
Thank very much you for your response. We have removed the sidecar container and NEURON_RTD_ADDRESS (previously set to be unix:/run/neuron.sock) environment variable from the deploy.yml file. 

I executed the following command on the EC2 node machine. However, we are seeing that the neuron-rtd cannot be stopped because it is not there in the first place. 

# Configure Linux for Neuron repository updates
sudo tee /etc/yum.repos.d/neuron.repo > /dev/null <<EOF
[neuron]
name=Neuron YUM Repository
baseurl=https://yum.repos.neuron.amazonaws.com
enabled=1
metadata_expire=0
EOF
sudo rpm --import https://yum.repos.neuron.amazonaws.com/GPG-PUB-KEY-AMAZON-AWS-NEURON.PUB
# Stop any existing Neuron Runtime 1.x daemon (neuron-rtd)
**sudo systemctl stop neuron-rtd** --> **Does not work because the system cannot find neuron-rtd**
# Uninstall neuron-rtd
sudo yum remove aws-neuron-runtime
# Update OS packages
sudo yum update -y
# Install OS headers
sudo yum install kernel-devel-$(uname -r) kernel-headers-$(uname -r) -y
# Remove existing Neuron Driver
sudo yum remove aws-neuron-dkms
# Install Neuron Driver
sudo yum install aws-neuron-dkms -y


After performing the steps above, we still somehow see the same error mentioned in the previous comment **"The Neuron Runtime GRPC server at address unix:/run/neuron.sock is unavailable. Is neuron-rtd running? Is socket /run/neuron.sock writable?"**. 


A couple of observations.

1) Since NEURON_RTD_ADDRESS is no longer defined, neuron-rtd should have been started within the same container according to this [post](https://github.com/aws/aws-neuron-sdk/issues/290#issuecomment-865109155). I'm not sure why it is not the case. 
2) How come the system is still looking for the neuron-rtd in the first place when I already upgraded aws-neuron-tools and aws-neuron-dkms. 

I attached relevant files here as well. Any clarification is greatly appreciated. 

**Dockerfile** 

# syntax = docker/dockerfile:experimental
#
# This file can build images for cpu and gpu env. By default it builds image for CPU.
# Use following option to build image for cuda/GPU: --build-arg BASE_IMAGE=nvidia/cuda:10.1-cudnn7-runtime-ubuntu18.04
# Here is complete command for GPU/cuda - 
# $ DOCKER_BUILDKIT=1 docker build --file Dockerfile --build-arg BASE_IMAGE=nvidia/cuda:10.1-cudnn7-runtime-ubuntu18.04 -t torchserve:latest .
#
# Following comments have been shamelessly copied from https://github.com/pytorch/pytorch/blob/master/Dockerfile
# 
# NOTE: To build this you will need a docker version > 18.06 with
#       experimental enabled and DOCKER_BUILDKIT=1
#
#       If you do not use buildkit you are not going to have a good time
#
#       For reference: 
#           https://docs.docker.com/develop/develop-images/build_enhancements/


ARG BASE_IMAGE=ubuntu:18.04

FROM ${BASE_IMAGE} AS compile-image

ENV PYTHONUNBUFFERED TRUE

RUN --mount=type=cache,id=apt-dev,target=/var/cache/apt \
    apt-get update && \
    DEBIAN_FRONTEND=noninteractive apt-get install --no-install-recommends -y \
    ca-certificates \
    g++ \
    python3-dev \
    python3-distutils \
    python3-venv \
    openjdk-11-jre-headless \
    curl \
    && rm -rf /var/lib/apt/lists/* \
    && cd /tmp \
    && curl -O https://bootstrap.pypa.io/get-pip.py \
    && python3 get-pip.py
   
RUN python3 -m venv /home/venv

ENV PATH="/home/venv/bin:$PATH"

RUN update-alternatives --install /usr/bin/python python /usr/bin/python3 1
RUN update-alternatives --install /usr/local/bin/pip pip /usr/local/bin/pip3 1

# This is only useful for cuda env 
RUN export USE_CUDA=1

RUN pip install --upgrade pip
RUN pip install --no-cache-dir --extra-index-url https://pip.repos.neuron.amazonaws.com torch-neuron==1.5.1.1.0.1721.0 torchvision torchtext torchserve torch-model-archiver regex numpy 
RUN pip install transformers==2.5.1 
RUN pip install --no-cache-dir --extra-index-url https://pip.repos.neuron.amazonaws.com neuron-cc[tensorflow]

# Final image for production

FROM ${BASE_IMAGE} AS runtime-image

ENV PYTHONUNBUFFERED TRUE

RUN --mount=type=cache,target=/var/cache/apt \
    apt-get update && \
    DEBIAN_FRONTEND=noninteractive apt-get install --no-install-recommends -y \
    python3 \
    python3-distutils \
    openjdk-11-jre-headless \
    ca-certificates \
    wget \
    gnupg2 \
    && rm -rf /var/lib/apt/lists/* \
    && cd /tmp

COPY neuron.list /etc/apt/sources.list.d/neuron.list
RUN wget -qO - https://apt.repos.neuron.amazonaws.com/GPG-PUB-KEY-AMAZON-AWS-NEURON.PUB | apt-key add -

RUN --mount=type=cache,target=/var/cache/apt \
    apt-get update && \
    DEBIAN_FRONTEND=noninteractive apt-get install --no-install-recommends -y \
    # aws-neuron-runtime-base \
    # aws-neuron-runtime \
    aws-neuron-tools

COPY --from=compile-image /home/venv /home/venv

ENV PATH="/home/venv/bin:$PATH"
ENV PATH="/opt/aws/neuron/bin:$PATH"

RUN useradd -m model-server && mkdir -p /home/model-server/tmp
RUN chown -R model-server /home/model-server
RUN mkdir /home/model-server/model-store && chown -R model-server /home/model-server/model-store

COPY dockerd-entrypoint.sh /usr/local/bin/dockerd-entrypoint.sh
RUN chmod +x /usr/local/bin/dockerd-entrypoint.sh

COPY liveness.sh /home/model-server/liveness.sh
RUN chmod +x /home/model-server/liveness.sh

COPY config.properties /home/model-server/config.properties
COPY log4j.properties /home/model-server/log4j.properties

EXPOSE 8080 8081

USER model-server
WORKDIR /home/model-server
ENV TEMP=/home/model-server/tmp
ENTRYPOINT ["/usr/local/bin/dockerd-entrypoint.sh"]

**dockerd-entrypoint.sh**
#!/bin/bash
set -e

if [[ "$1" = "serve" ]]; then
    shift 1
    torchserve --start --ncs --ts-config /home/model-server/config.properties --log-config /home/model-server/log4j.properties --model-store /home/model-server/model-store "$@"
else
    eval "$@"
fi

# prevent docker exit
tail -f /dev/null

**Selected portion of deploy.yaml**
apiVersion: v1
kind: Service
metadata:
  name: xxx
  labels:
    app: xxx
spec:
  selector:
    app: xxx
  ports:
    - port: 80
      targetPort: 8080
      protocol: TCP
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: xxx
spec:
  strategy:
    type: RollingUpdate
    rollingUpdate:
      maxUnavailable: 1
      maxSurge: 0
  selector:
    matchLabels:
      app: xxx
  template:
    metadata:
      labels:
        app: xxx
      annotations:
        linkerd.io/inject: enabled
    spec:
      volumes:
        - name: sock
          emptyDir: {}
        - name: model-file
          hostPath:
            path: /opt/models/model-store
            type: Directory
      nodeSelector:
        workertype: xxx
      containers:
        - name: xxx
          image: xxx
          ports:
            - containerPort: 8080
          imagePullPolicy: IfNotPresent
          args:
            - serve
            - --models=xxx.mar
          livenessProbe:
            exec:
              command:
              - /bin/sh
              - -c
              - /home/model-server/liveness.sh
            initialDelaySeconds: 60
            periodSeconds: 15
            failureThreshold: 1
          readinessProbe:
            httpGet:
              path: /ping
              port: 8080
            initialDelaySeconds: 45
            periodSeconds: 5

          volumeMounts:
            - name: model-file
              mountPath: "/home/model-server/model-store"

          resources:
            limits:
              hugepages-2Mi: 256Mi    # configure to 256 * desired number of Inferentia devices.
              aws.amazon.com/neuron: 1  # desired number of Inferentia devices.
            requests:
              memory: 1024Mi

          securityContext:
            capabilities:
              add:
                - IPC_LOCK

**Comment by micwade-aws** [comment]
Thanks - the docker files were helpful.  

> 2. How come the system is still looking for the neuron-rtd in the first place..?

The version of torch-neuron being installed ([1.5.1.1.0.1721.0](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/release-notes/conda/torch/conda-torch-neuron.html?highlight=1.5.1.1.0.1721.0#id21)) is an older version of pytorch from last year that does not use the integrated runtime (libnrt)

From your dockerfile:
```
RUN pip install --no-cache-dir --extra-index-url https://pip.repos.neuron.amazonaws.com torch-neuron==1.5.1.1.0.1721.0 torchvision torchtext torchserve torch-model-archiver regex numpy 
````

Is it possible to use torch-neuron-1.5.1.2.0.392.0 instead, which is the latest in [Neuron 1.16.1](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/release-notes/releasecontent.html#neuron-release-content)?

**Comment by awsrjh** [comment]
Hi - has this helped resolve this issue?

**Comment by ndenStanford** [comment]
It works at least on my testing cluster. Thank you so much for asking. I truly appreciate the team's help!
